plugins/hls-rename-plugin/src/Ide/Plugin/Rename.hs

```diff
@@ -57,43 +58,66 @@ import           Language.LSP.Server
 instance Hashable (Mod a) where hash n = hash (unMod n)
 
 descriptor :: Recorder (WithPriority E.Log) -> PluginId -> PluginDescriptor IdeState
-descriptor recorder pluginId = mkExactprintPluginDescriptor recorder $ (defaultPluginDescriptor pluginId "Provides renaming of Haskell identifiers")
-    { pluginHandlers = mkPluginHandler SMethod_TextDocumentRename renameProvider
-    , pluginConfigDescriptor = defaultConfigDescriptor
-        { configCustomConfig = mkCustomConfig properties }
-    }
+descriptor recorder pluginId = mkExactprintPluginDescriptor recorder $
+    (defaultPluginDescriptor pluginId "Provides renaming of Haskell identifiers")
+        { pluginHandlers = mconcat
+              [ mkPluginHandler SMethod_TextDocumentRename renameProvider
+              , mkPluginHandler SMethod_TextDocumentPrepareRename prepareRenameProvider
+              ]
+        , pluginConfigDescriptor = defaultConfigDescriptor
+            { configCustomConfig = mkCustomConfig properties }
+        }
+
+prepareRenameProvider :: PluginMethodHandler IdeState Method_TextDocumentPrepareRename
+prepareRenameProvider state _pluginId (PrepareRenameParams (TextDocumentIdentifier uri) pos _progressToken) = do
+    nfp <- getNormalizedFilePathE uri
+    namesUnderCursor <- getNamesAtPos state nfp pos
+    -- When this handler says that rename is invalid, VSCode shows "The element can't be renamed"
+    -- and doesn't even allow you to create full rename request.
+    -- This handler deliberately approximates "things that definitely can't be renamed"
+    -- to mean "there is no Name at given position".
+    --
+    -- In particular it allows some cases through (e.g. cross-module renames),
+    -- so that the full rename handler can give more informative error about them.
+    let renameValid = not $ null namesUnderCursor
+    pure $ InL $ PrepareRenameResult $ InR $ InR $ #defaultBehavior .== renameValid
```

Comment:
I'm unsure if this is the correct behaviour. It's very hard to tell what the intention of the various options for results of `prepareRename` are. Worth a comment anyway. We could consider returning the range corresponding to the name, although it's unclear if we have that...

Reply:
I would say that [lsp spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_prepareRename) is quite clear in this regard:

>  If { defaultBehavior: boolean } is returned (since 3.16) the rename position is valid and the client should use its default behavior to compute the rename range.

I looked at what the vscode-languageclient is doing with that response (see [here](https://github.com/microsoft/vscode-languageserver-node/blob/46f6f9038fee60dbdfeb5307201397adc8feb3cb/client/src/common/rename.ts#L94-L108)) and when you return the boolean it does:
if true it triggers the default rename behavior (open input box where you can specify the new name)
if false it gives you a small error popup saying "The element can't be renamed"

I'll let you read this / react and will can open a PR with some additional clarifying comment later today.

Reply:
I guess I'm just wondering if we can rely on the "default behaviour" or whether it would be better for us to be returning the actual range of the name? Maybe it's fine?

Reply:
I played with it for a bit and it seems it would be hard to get this right if we're to provide the Range under cursor.

The thing I had trouble to make work was qualified names. 
e.g. if you rename `IO.putStrLn` the HieAST contains the range of the whole thing (including `IO.`) so if you trigger rename on it, you get the whole thing prefilled into the rename input which is confusing.

I'd be more comfortable keeping the current implementation which seems to work robustly on all cases we have in tests.

Reply:
Huh, and vscode does the right thing and just changes `putStrLn`? IDK, seems potentially fragile but maybe it's fine.

Anyway, this all seems like super useful information, so worth writing down!